File: ConstructBinaryTreeFromPreorderAndPostorder.py
#Time Complexity - O(n)
#Space Complexity - O(Height of the tree)
# Problem successfully executed on leetcode
# Had to take min and max NULL at first in order to incorporate the logic for edge cases nodes such as INT_MIN and INT_MAX values.

class TreeNode:
     def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if preorder== None or inorder==None or len(preorder)==0:
            return None
        root=TreeNode(preorder[0])
        mid=inorder.index(root.val)
        
        root.left=self.buildTree(preorder[1:mid+1],inorder[:mid])
        root.right=self.buildTree(preorder[mid+1:],inorder[mid+1:])
        return root
        
        
        
        
        
        

File: Construct binary tree from preorder and postorder.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
// Your code here along with comments explaining your approach: Since the first element of the preorder is always the root, we will look for that root in the inorder array using a hashmap. Then we will set that as the rootIdx. 
//This will help us set the range for the left and the right subtree. The range for left subtree is from the start to the rootIdx-1 and for right subtree is rootIdx+1 to the end. 


class Solution {
    int idx;
    HashMap<Integer,Integer> map;

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        this.map=new HashMap<>();
        //Storing the values of Inorder in HashMap

        for(int i=0;i<inorder.length;i++)
        {
            map.put(inorder[i],i);
        }
        return helper(preorder,0,inorder.length-1,0);

    }

    private TreeNode helper(int[] preorder, int start, int end,int rootIdx)
  {  //baseif 
    if(start>end) return null;
    //logic
    int rootval=preorder[idx];
    idx++;
    //search for the rootval in inorder
    rootIdx=map.get(rootval);
    //Construct the tree
    TreeNode root=new TreeNode(rootval);

    root.left=helper(preorder,start,rootIdx-1,rootIdx);
    root.right=helper(preorder,rootIdx+1,end,rootIdx);

    return root;
}


}
